src/ray/util/process.cc

```diff
@@ -50,60 +53,112 @@ class ProcessFD {
   pid_t GetId() const;
 
   // Fork + exec combo. Returns -1 for the PID on failure.
-  static ProcessFD spawnvp(const char *argv[], std::error_code &ec) {
+  static ProcessFD spawnvp(const char *argv[], std::error_code &ec, bool decouple) {
     ec = std::error_code();
     intptr_t fd;
     pid_t pid;
 #ifdef _WIN32
+    (void)decouple;  // Windows doesn't require anything particular for decoupling.
     std::vector<std::string> args;
     for (size_t i = 0; argv[i]; ++i) {
       args.push_back(argv[i]);
     }
-    // Calling CreateCommandLine() here wouldn't make sense here if the
-    // Microsoft C runtime properly quoted each command-argument argument.
-    // However, it doesn't quote at all. It just joins arguments with a space.
-    // So we have to do the quoting manually and pass everything as a single argument.
-    fd = _spawnlp(P_NOWAIT, args[0].c_str(), CreateCommandLine(args).c_str(), NULL);
-    if (fd != -1) {
-      pid = static_cast<pid_t>(GetProcessId(reinterpret_cast<HANDLE>(fd)));
-      if (pid == 0) {
-        pid = -1;
+    std::string cmds[] = {std::string(), CreateCommandLine(args)};
+    if (GetFileName(args.at(0)).find('.') == std::string::npos) {
+      // Some executables might be missing an extension.
+      // Append a single "." to prevent automatic appending of extensions by the system.
+      std::vector<std::string> args_direct_call = args;
+      args_direct_call[0] += ".";
+      cmds[0] = CreateCommandLine(args_direct_call);
+    }
+    bool succeeded = false;
+    PROCESS_INFORMATION pi = {};
+    for (int attempt = 0; attempt < sizeof(cmds) / sizeof(*cmds); ++attempt) {
+      std::string &cmd = cmds[attempt];
+      if (!cmd.empty()) {
+        (void)cmd.c_str();  // We'll need this to be null-terminated (but mutable) below
```

Comment:
Hmm, not sure if the C++ standard allows this to be mutable. It's probably ok, just wanted to point it out.

Reply:
Yeah, I actually mostly kept this line for documentation than anything else. In reality implementations often null-terminated the string a priori, so this ends up being like `data()`. But I think it's technically possible for an implementation to only null-terminate when `c_str()` is called, so I added it to do it explicitly.